arangod/Aql/QueryRegistry.cpp

```diff
@@ -182,146 +188,201 @@ void* QueryRegistry::openEngine(EngineId id, EngineType type) {
-  }
+  {
+    WRITE_LOCKER(writeLocker, _lock);
 
-  ei._isOpen = false;
+    auto it = _engines.find(engineId);
+    if (it == _engines.end()) {
+      LOG_TOPIC("97351", DEBUG, arangodb::Logger::AQL)
+          << "Found no engine with id " << engineId;
+      return;
+    }
 
-  if (ei._queryInfo) {
-    TRI_ASSERT(ei._queryInfo->_numOpen > 0);
-    ei._queryInfo->_numOpen--;
-    if (!ei._queryInfo->_query->killed() && ei._queryInfo->_expires != 0) {
-      ei._queryInfo->_expires = TRI_microtime() + ei._queryInfo->_timeToLive;
+    EngineInfo& ei = it->second;
+    if (!ei._isOpen) {
+      LOG_TOPIC("45b8e", DEBUG, arangodb::Logger::AQL)
+          << "engine id " << engineId << " was not open.";
+      THROW_ARANGO_EXCEPTION_MESSAGE(
+          TRI_ERROR_INTERNAL, "engine with given vocbase and id is not open");
+    }
+
+    ei._isOpen = false;
+
+    if (ei._queryInfo) {
+      TRI_ASSERT(ei._queryInfo->_numOpen > 0);
+      ei._queryInfo->_numOpen--;
+      if (ei._queryInfo->_numOpen == 0 && ei._queryInfo->_finished) {
+        // we were the last thread to close the engine, but the query has
+        // already been marked as finished, so we are responsible to resolve the
+        // future in order to actually finish the query
+        queryInfoToFinish = ei._queryInfo;
+      }
+      if (!ei._queryInfo->_query->killed() && ei._queryInfo->_expires != 0) {
+        ei._queryInfo->_expires = TRI_microtime() + ei._queryInfo->_timeToLive;
+      }
+      LOG_TOPIC("5ecdc", TRACE, arangodb::Logger::AQL)
+          << "closing engine " << engineId
+          << ", query id: " << ei._queryInfo->_query->id()
+          << ", numOpen: " << ei._queryInfo->_numOpen;
+    } else {
+      LOG_TOPIC("ae981", TRACE, arangodb::Logger::AQL)
+          << "closing engine " << engineId << ", no query";
     }
-    LOG_TOPIC("5ecdc", TRACE, arangodb::Logger::AQL)
-        << "closing engine " << engineId
-        << ", query id: " << ei._queryInfo->_query->id()
-        << ", numOpen: " << ei._queryInfo->_numOpen;
-  } else {
-    LOG_TOPIC("ae981", TRACE, arangodb::Logger::AQL)
-        << "closing engine " << engineId << ", no query";
+  }
+
+  if (queryInfoToFinish) {
+    queryInfoToFinish->_promise.setValue(queryInfoToFinish->_query);
   }
 }
 
 /// @brief destroy
 // cppcheck-suppress virtualCallInConstructor
-std::shared_ptr<ClusterQuery> QueryRegistry::destroyQuery(
-    std::string const& vocbase, QueryId id, ErrorCode errorCode) {
-  std::unique_ptr<QueryInfo> queryInfo;
+void QueryRegistry::destroyQuery(std::string const& vocbase, QueryId id,
+                                 ErrorCode errorCode) {
+  WRITE_LOCKER(writeLocker, _lock);
 
-  {
-    WRITE_LOCKER(writeLocker, _lock);
+  QueryInfoMapPerVocbase::iterator vocbaseQueriesIt;
+  QueryInfoMap::iterator queryMapIt;
+  if (!lookupQueryForFinalization(vocbase, id, errorCode, vocbaseQueriesIt,
+                                  queryMapIt)) {
+    return;
+  }
 
```

Comment:
I think we should add assertions here that both `vocbaseQueriesIt` and `queryMapIt` are populated as expected and are valid iterators.